vendor/knative.dev/pkg/client/injection/kube/informers/factory/filtered/fake/fake_filtered_factory.go

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Code generated by injection-gen. DO NOT EDIT.
+
+package fakeFilteredFactory
+
+import (
+	context "context"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	informers "k8s.io/client-go/informers"
+	fake "knative.dev/pkg/client/injection/kube/client/fake"
+	filtered "knative.dev/pkg/client/injection/kube/informers/factory/filtered"
+	controller "knative.dev/pkg/controller"
+	injection "knative.dev/pkg/injection"
+	logging "knative.dev/pkg/logging"
+)
+
+var Get = filtered.Get
+
+func init() {
+	injection.Fake.RegisterInformerFactory(withInformerFactory)
+}
+
+func withInformerFactory(ctx context.Context) context.Context {
+	c := fake.Get(ctx)
+	opts := []informers.SharedInformerOption{}
+	if injection.HasNamespaceScope(ctx) {
+		opts = append(opts, informers.WithNamespace(injection.GetNamespaceScope(ctx)))
+	}
+	untyped := ctx.Value(filtered.LabelKey{})
+	if untyped == nil {
+		logging.FromContext(ctx).Panic(
+			"Unable to fetch labelkey from context.")
+	}
+	labelSelectors := untyped.([]string)
+	for _, selector := range labelSelectors {
+		thisOpts := append(opts, informers.WithTweakListOptions(func(l *v1.ListOptions) {
+			l.LabelSelector = selector
```

Comment:
I know this is in the vendored pkg which already landed, but seeing it in context I think this is wrong: if multiple selectors are passed each one will override the single labelSelector field, so only the last selector will have any effect, I think?

Reply:
(see https://github.com/knative/pkg/blob/b558677ab03404ed118ba3e179a7438fe2d8509a/vendor/k8s.io/client-go/informers/factory.go#L80-L85, looks like WithTweakListOptions isn't safe to pass multiple times in one options list, unfortunately)

Reply:
@julz ,  After re-reading the code in details,    I found it should not be an issue. 

If we have multiple labelSelectors as` []string{"key1=value1", "key2=value2"},  then,  the below code will add  2 SharedInformerFactory instances into ctx. 
*  SharedInformerFactory instance 1 with `opt  selector "key1=value1"` , then add to ctx.WithValue(ctx, key{Selector:"key1=value1"}, ...)
*  SharedInformerFactory instance 2 with `opt  selector "key2=value2"` , then add to ctx.WithValue(ctx, key{Selector:"key2=value2"}, ...)

```
	labelSelectors := untyped.([]string)
	for _, selector := range labelSelectors {
		thisOpts := append(opts, informers.WithTweakListOptions(func(l *v1.ListOptions) {
			l.LabelSelector = selector
		}))
		ctx = context.WithValue(ctx, Key{Selector: selector},
			informers.NewSharedInformerFactoryWithOptions(c, controller.GetResyncPeriod(ctx), thisOpts...))
	}
```

Then, later on , with Get(ctx, "key1=value1"),  only the SharedInformerFactory instance 1 will be returned and used in further. 

As a result, the selector `key1=value1` is not override by `key2=value2`, and no multiple selectors are passed to WithTweakListOptions().

So, I think the current code in pkg is fine .  Can you double check? 

> I know this is in the vendored pkg which already landed, but seeing it in context I think this is wrong: if multiple selectors are passed each one will override the single labelSelector field, so only the last selector will have any effect, I think?

Reply:
I think this may depend, unfortunately, on how `append` chooses to add to the array. Since `append` works in-place it is modifying the underlying `opts` array each time, which means it's possible for the first selector to be clobbered by the second, I think, depending on the capacity of the array. Here's an example showing what Im worried about - one append can clobber the other even if assigned to a new var each time through the loop - https://play.golang.org/p/semtv5-M-xv.

I'd suggest we either simplify WithSelectors to just WithSelector (since there doesn't seem to be any actual use case for WithSelectors here..), or else move creation of thisOpt all in to the loop so we don't have to worry about odd append semantics :).

Reply:
@julz , for https://play.golang.org/p/semtv5-M-xv,  yes, the append "b" will overwrite "a" since the underlyarray cap is 5.  But in the current implementation, the `opts` is claimed with len=0 & cap=0, so each append will create a new underly array , and won't override. 
``` 
	opts := []informers.SharedInformerOption{}
```

see https://play.golang.org/p/DtTHHxSFLVc

Also, I checked the history and noted the current implementation was based on  @mattmoor 's comment https://github.com/knative/pkg/pull/1940#issuecomment-739996222

Just want to confirm whether the support for above comment is not necessary now?